aten/src/ATen/cuda/CublasHandlePool.cpp

```diff
@@ -41,6 +41,13 @@ cublasHandle_t getCurrentCUDABlasHandle() {
   auto handle = myPoolWindow->reserve(device);
   auto stream = c10::cuda::getCurrentCUDAStream();
   TORCH_CUDABLAS_CHECK(cublasSetStream(handle, stream));
+#if CUDA_VERSION >= 11000
+  if (at::globalContext().useTF32CuBLAS()) {
+    TORCH_CUDABLAS_CHECK(cublasSetMathMode(handle, CUBLAS_TF32_TENSOR_OP_MATH));
+  } else {
+    TORCH_CUDABLAS_CHECK(cublasSetMathMode(handle, CUBLAS_DEFAULT_MATH));
```

Comment:
Uh oh,  I don't think `CUBLAS_DEFAULT_MATH` is what we want here.  According to the [docs](https://docs.nvidia.com/cuda/cublas/index.html#cublasmath_t) it means 
> Prevent the library from using Tensor Core operations

We still want to use Tensor Cores for fp16 inputs, we just want to avoid tf32 arithmetic for fp32 inputs.

Reply:
On second thought, it's ok if downstream gemm ops that use the handle reset the math mode to `CUBLAS_TENSOR_OP_MATH` when they detect fp16 inputs.

Reply:
I believe there is some code turning it on on gemm:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDABlas.cpp#L206
```C++
  if (prop->major >= 5) {
    TORCH_CUDABLAS_CHECK(cublasSetMathMode(handle, CUBLAS_TENSOR_OP_MATH));
    TORCH_CUDABLAS_CHECK(cublasGemmEx(
```
Also, from cuBLAS documentation:

>The math mode default is CUBLAS_DEFAULT_MATH, this default indicates that the Tensor Core operations will be avoided by the library. The default mode is a serialized operation, the Tensor Core operations are parallelized, thus the two might result in slight different numerical results due to the different sequencing of operations.

Reply:
heh, i think you were typing that while i was looking for the downstream ops and editing my comment.

Reply:
I thought of something else, less pressing:  Ampere has fp64 tensor cores!  We'd also like fp64 gemms to use those, scientific computing users will love you.  Maybe [`gemm<double>`](gemm<double>) also needs
```
  if (prop->major >= 8) {
    TORCH_CUDABLAS_CHECK(cublasSetMathMode(handle, MODE_FOR_FP64_TENSOR_CORES));
```
However, Ampere docs aren't out yet, so I have no idea what `MODE_FOR_FP64_TENSOR_CORES` might be.  Maybe it will also be `CUBLAS_TENSOR_OP_MATH`.  In any case I think fp64 tensor core support can be added in a later PR.

Reply:
I like the idea:
> While we're at it, if setting cublasSetMathMode(handle, WHATEVER); then resetting it at the end of a function becomes a common pattern, we should make a dedicated RAII guard.

Reply:
After discussions with Cublas people I think the only additional changes we need for future-proofing are guarding `cublasSetMathMode` with `#if CUDA_VERSION < 11000` [here](https://github.com/pytorch/pytorch/blob/ba0851326c87256a819197c133b44f3e617b0608/aten/src/ATen/cuda/CUDABlas.cpp#L206), [here](https://github.com/pytorch/pytorch/blob/ba0851326c87256a819197c133b44f3e617b0608/aten/src/ATen/cuda/CUDABlas.cpp#L227), [here](https://github.com/pytorch/pytorch/blob/d7f7c290e3d76a1e3019166644baf78de0d95a31/aten/src/THC/THCBlas.cu#L274), [here](https://github.com/pytorch/pytorch/blob/d7f7c290e3d76a1e3019166644baf78de0d95a31/aten/src/THC/THCBlas.cu#L281).  `CUBLAS_TF32_TENSOR_OP_MATH` will allow tensor cores for fp16 and fp64, we don't need to manually set `CUBLAS_TENSOR_OP_MATH`.

Reply:
Would an RAII guard that mindlessly restore the original flag makes more sense? This is mainly for better code readability, instead of asking code reader to think about the logic on which flag has which behavior, an RAII guard that restores the original flag requires no thinking and therefore more maintainable.

Reply:
By "original flag" do you mean the surrounding mode, or `CUBLAS_MATH_DEFAULT` no matter what the surrounding mode used to be?  I definitely think a guard should save and restore surrounding state, not silently alter it on exit to something the gemm caller doesn't expect.

I don't think a guard is absolutely necessary here.  I think adding the macros at those four spots is necessary and sufficient, and a guard is maybe a useful alternative.  Do you agree the macros would get us [desired behavior for all cases](https://nvidia.slack.com/archives/C8HLVCCUV/p1591046650180700?thread_ts=1591045401.177500&cid=C8HLVCCUV), with the possibility of a two-gemm op in mind?

Reply:
Yeah, I agree now. Basically what happened on CUDA 11 is: the flag is set when a handle is obtained, and never modified later. And before CUDA 11, the flag starts from the default, gets flipped when using FP16, and flipped back to default when done. I am now convinced that using a CUDA 11 guard to disable all flag changes is a better approach than the RAII guard. Thanks for the discussion!

Reply:
hmm a guard would still need macros at the point of use, because using the guard to set `CUBLAS_TENSOR_OP_MATH` should be elided for cuda>=11, unless we provide our own enum flag layer e.g. `DO/DONT_ALLOW_TENSOR_OPS` that the guard's constructor accepts, and the constructor contains internal `#if CUDA_VERSION < 11000...else....` branches that translate it to the right version-dependent cublas flag.

It's really up to you, as long as we get the right behavior at the end of the day.

Reply:
I believe this is resolved.